daemon/create_windows.go

```diff
@@ -80,3 +85,67 @@ func (daemon *Daemon) createContainerOSSpecificSettings(container *container.Con
 	}
 	return nil
 }
+
+// checkImageIsolationCompatiblity checks whether the provided image is compatible
+// with the host config's isolation settings.
+func (daemon *Daemon) checkImageIsolationCompatiblity(hostConfig *containertypes.HostConfig, img *image.Image) error {
+	hostOSV := osversion.Get()
+	isHyperV := hostConfig.Isolation.IsHyperV()
+	// fallback on default daemon isolation if none was explicitly provided:
+	if hostConfig.Isolation.IsDefault() {
+		isHyperV = daemon.defaultIsolation.IsHyperV()
+	}
+	return checkImageCompatibilityForHostIsolation(hostOSV, img.OSVersion, isHyperV)
+}
+
+// checkImageCompatibilityForHostIsolation checks whether the provided `imageOSVersion`
+// should be able to run on the provided host OS version given Hyper-V isolation.
+// Its contained logic can be distilled into:
+// - if imageOS == hostOS => can run under any isolation mode
+// - if imageOS < hostOS => must be run under Hyper-V
+// - if imageOS > hostOS => can be run under Hyper-V only on an RS5 (1809+) host
+// Please see the below:
+// https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2016%2Cwindows-10#windows-server-host-os-compatibility
+func checkImageCompatibilityForHostIsolation(hostOSV osversion.OSVersion, imageOSVersion string, isHyperV bool) error {
+	var err error
+	var imageOSBuild int
+	splitImageOSVersion := strings.Split(imageOSVersion, ".") // eg 10.0.16299.nnnn
+	if len(splitImageOSVersion) >= 3 {
+		if imageOSBuild, err = strconv.Atoi(splitImageOSVersion[2]); err == nil {
+			logrus.Debugf("parsed Windows build number %d from image OS version %s", imageOSBuild, imageOSVersion)
+		} else if imageOSBuild < 0 {
```

Comment:
This case will only ever be evaluated if `err != nil`, so we don't really need it. If we are worried about negative numbers in the version string, suggest to use `strconv.ParseUint` instead of `strconv.Atoi`.